go/cmd/dolt/commands/stashcmds/stash.go

```diff
@@ -0,0 +1,174 @@
+	StashClearCmd{},
+	StashDropCmd{},
+	StashListCmd{},
+	StashPopCmd{},
+})
+
+var stashDocs = cli.CommandDocumentationContent{
+	ShortDesc: "Stash the changes in a dirty working directory away.",
+	LongDesc: `Use dolt stash when you want to record the current state of the working directory and the index, but want to go back to a clean working directory. 
+
+The command saves your local modifications away and reverts the working directory to match the HEAD commit.
+`,
+	Synopsis: []string{
+		"list",
+		"pop {{.LessThan}}stash{{.GreaterThan}}",
+		"clear",
+		"drop {{.LessThan}}stash{{.GreaterThan}}",
+	},
+}
+
+type StashCmd struct{}
+
+// Name is returns the name of the Dolt cli command. This is what is used on the command line to invoke the command
+func (cmd StashCmd) Name() string {
+	return "stash"
+}
+
+// Description returns a description of the command
+func (cmd StashCmd) Description() string {
+	return "Stash the changes in a dirty working directory away."
+}
+
+func (cmd StashCmd) Docs() *cli.CommandDocumentation {
+	ap := cmd.ArgParser()
+	return cli.NewCommandDocumentation(stashDocs, ap)
+}
+
+func (cmd StashCmd) ArgParser() *argparser.ArgParser {
+	ap := argparser.NewArgParser()
+	return ap
+}
+
+// EventType returns the type of the event to log
+func (cmd StashCmd) EventType() eventsapi.ClientEventType {
+	return eventsapi.ClientEventType_STASH
+}
+
+// Exec executes the command
+func (cmd StashCmd) Exec(ctx context.Context, commandStr string, args []string, dEnv *env.DoltEnv) int {
+	if !dEnv.DoltDB.Format().UsesFlatbuffers() {
+		cli.PrintErrln(ErrStashNotSupportedForOldFormat.Error())
+		return 1
+	}
+	ap := cmd.ArgParser()
+	help, usage := cli.HelpAndUsagePrinters(cli.CommandDocsForCommandString(commandStr, stashDocs, ap))
+	apr := cli.ParseArgsOrDie(ap, args, help)
+	if dEnv.IsLocked() {
+		return commands.HandleVErrAndExitCode(errhand.VerboseErrorFromError(env.ErrActiveServerLock.New(dEnv.LockFile())), help)
+	}
+
+	if apr.NArg() > 0 {
+		usage()
+		return 1
+	}
+
+	err := stashChanges(ctx, dEnv)
+	if err != nil {
+		return commands.HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usage)
+	}
+	return 0
+}
+
+func stashChanges(ctx context.Context, dEnv *env.DoltEnv) error {
+	headRoot, err := dEnv.HeadRoot(ctx)
+	if err != nil {
+		return err
+	}
+	workingRoot, err := dEnv.WorkingRoot(ctx)
+	if err != nil {
+		return err
+	}
+	stagedRoot, err := dEnv.StagedRoot(ctx)
+	if err != nil {
+		return err
+	}
+	headHash, err := headRoot.HashOf()
+	if err != nil {
+		return err
+	}
+	workingHash, err := workingRoot.HashOf()
+	if err != nil {
+		return err
+	}
+	stagedHash, err := stagedRoot.HashOf()
+	if err != nil {
+		return err
+	}
+
+	if headHash.Equal(workingHash) && headHash.Equal(stagedHash) {
+		return fmt.Errorf("No local changes to save")
+	}
+
+	// TODO: handle cases with staged changes?
```

Comment:
Agreed to get a separate PR out for handling staged changes. For now, having staged set of changes will not allow user to stash their changes. It will be fixed soon.